# Part I - Ames Data Housing)
## by (David Babayo)

## Introduction
> This data set includes information about houses with multiple features with the important feature being "SalePrice" which we would at the end of this we would have models which we would use to predict this particular feature based on the house features.


## Preliminary Wrangling

In [2]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

In [5]:
df = pd.read_csv("Ames_Housing_Data.csv")

In [6]:
df.head()

PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley  \
0  526301100           20        RL         141.0     31770   Pave   NaN   
1  526350040           20        RH          80.0     11622   Pave   NaN   
2  526351010           20        RL          81.0     14267   Pave   NaN   
3  526353030           20        RL          93.0     11160   Pave   NaN   
4  527105010           60        RL          74.0     13830   Pave   NaN   

  Lot Shape Land Contour Utilities  ... Pool Area Pool QC  Fence Misc Feature  \
0       IR1          Lvl    AllPub  ...         0     NaN    NaN          NaN   
1       Reg          Lvl    AllPub  ...         0     NaN  MnPrv          NaN   
2       IR1          Lvl    AllPub  ...         0     NaN    NaN         Gar2   
3       Reg          Lvl    AllPub  ...         0     NaN    NaN          NaN   
4       IR1          Lvl    AllPub  ...         0     NaN  MnPrv          NaN   

  Misc Val Mo Sold Yr Sold  Sale Type  Sale Condition  SalePrice  
0        0       5    2010        WD           Normal     215000  
1        0       6    2010        WD           Normal     105000  
2    12500       6    2010        WD           Normal     172000  
3        0       4    2010        WD           Normal     244000  
4        0       3    2010        WD           Normal     189900  

[5 rows x 81 columns]

In [7]:
df.shape

(2930, 81)

In [8]:
df.dtypes

PID                 int64
MS SubClass         int64
MS Zoning          object
Lot Frontage      float64
Lot Area            int64
                   ...   
Mo Sold             int64
Yr Sold             int64
Sale Type          object
Sale Condition     object
SalePrice           int64
Length: 81, dtype: object

In [9]:
df.describe().transpose()

count          mean           std          min  \
PID              2930.0  7.144645e+08  1.887308e+08  526301100.0   
MS SubClass      2930.0  5.738737e+01  4.263802e+01         20.0   
Lot Frontage     2440.0  6.922459e+01  2.336533e+01         21.0   
Lot Area         2930.0  1.014792e+04  7.880018e+03       1300.0   
Overall Qual     2930.0  6.094881e+00  1.411026e+00          1.0   
Overall Cond     2930.0  5.563140e+00  1.111537e+00          1.0   
Year Built       2930.0  1.971356e+03  3.024536e+01       1872.0   
Year Remod/Add   2930.0  1.984267e+03  2.086029e+01       1950.0   
Mas Vnr Area     2907.0  1.018968e+02  1.791126e+02          0.0   
BsmtFin SF 1     2929.0  4.426296e+02  4.555908e+02          0.0   
BsmtFin SF 2     2929.0  4.972243e+01  1.691685e+02          0.0   
Bsmt Unf SF      2929.0  5.592625e+02  4.394942e+02          0.0   
Total Bsmt SF    2929.0  1.051615e+03  4.406151e+02          0.0   
1st Flr SF       2930.0  1.159558e+03  3.918909e+02        334.0   
2nd Flr SF       2930.0  3.354560e+02  4.283957e+02          0.0   
Low Qual Fin SF  2930.0  4.676792e+00  4.631051e+01          0.0   
Gr Liv Area      2930.0  1.499690e+03  5.055089e+02        334.0   
Bsmt Full Bath   2928.0  4.313525e-01  5.248202e-01          0.0   
Bsmt Half Bath   2928.0  6.113388e-02  2.452536e-01          0.0   
Full Bath        2930.0  1.566553e+00  5.529406e-01          0.0   
Half Bath        2930.0  3.795222e-01  5.026293e-01          0.0   
Bedroom AbvGr    2930.0  2.854266e+00  8.277311e-01          0.0   
Kitchen AbvGr    2930.0  1.044369e+00  2.140762e-01          0.0   
TotRms AbvGrd    2930.0  6.443003e+00  1.572964e+00          2.0   
Fireplaces       2930.0  5.993174e-01  6.479209e-01          0.0   
Garage Yr Blt    2771.0  1.978132e+03  2.552841e+01       1895.0   
Garage Cars      2929.0  1.766815e+00  7.605664e-01          0.0   
Garage Area      2929.0  4.728197e+02  2.150465e+02          0.0   
Wood Deck SF     2930.0  9.375188e+01  1.263616e+02          0.0   
Open Porch SF    2930.0  4.753345e+01  6.748340e+01          0.0   
Enclosed Porch   2930.0  2.301160e+01  6.413906e+01          0.0   
3Ssn Porch       2930.0  2.592491e+00  2.514133e+01          0.0   
Screen Porch     2930.0  1.600205e+01  5.608737e+01          0.0   
Pool Area        2930.0  2.243345e+00  3.559718e+01          0.0   
Misc Val         2930.0  5.063515e+01  5.663443e+02          0.0   
Mo Sold          2930.0  6.216041e+00  2.714492e+00          1.0   
Yr Sold          2930.0  2.007790e+03  1.316613e+00       2006.0   
SalePrice        2930.0  1.807961e+05  7.988669e+04      12789.0   

                          25%          50%           75%           max  
PID              5.284770e+08  535453620.0  9.071811e+08  1.007100e+09  
MS SubClass      2.000000e+01         50.0  7.000000e+01  1.900000e+02  
Lot Frontage     5.800000e+01         68.0  8.000000e+01  3.130000e+02  
Lot Area         7.440250e+03       9436.5  1.155525e+04  2.152450e+05  
Overall Qual     5.000000e+00          6.0  7.000000e+00  1.000000e+01  
Overall Cond     5.000000e+00          5.0  6.000000e+00  9.000000e+00  
Year Built       1.954000e+03       1973.0  2.001000e+03  2.010000e+03  
Year Remod/Add   1.965000e+03       1993.0  2.004000e+03  2.010000e+03  
Mas Vnr Area     0.000000e+00          0.0  1.640000e+02  1.600000e+03  
BsmtFin SF 1     0.000000e+00        370.0  7.340000e+02  5.644000e+03  
BsmtFin SF 2     0.000000e+00          0.0  0.000000e+00  1.526000e+03  
Bsmt Unf SF      2.190000e+02        466.0  8.020000e+02  2.336000e+03  
Total Bsmt SF    7.930000e+02        990.0  1.302000e+03  6.110000e+03  
1st Flr SF       8.762500e+02       1084.0  1.384000e+03  5.095000e+03  
2nd Flr SF       0.000000e+00          0.0  7.037500e+02  2.065000e+03  
Low Qual Fin SF  0.000000e+00          0.0  0.000000e+00  1.064000e+03  
Gr Liv Area      1.126000e+03       1442.0  1.742750e+03  5.642000e+03  
Bsmt Full Bath   0.000000e+00          0.0  1.00000

In [10]:
df.columns

Index(['PID', 'MS SubClass', 'MS Zoning', 'Lot Frontage', 'Lot Area', 'Street',
       'Alley', 'Lot Shape', 'Land Contour', 'Utilities', 'Lot Config',
       'Land Slope', 'Neighborhood', 'Condition 1', 'Condition 2', 'Bldg Type',
       'House Style', 'Overall Qual', 'Overall Cond', 'Year Built',
       'Year Remod/Add', 'Roof Style', 'Roof Matl', 'Exterior 1st',
       'Exterior 2nd', 'Mas Vnr Type', 'Mas Vnr Area', 'Exter Qual',
       'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure',
       'BsmtFin Type 1', 'BsmtFin SF 1', 'BsmtFin Type 2', 'BsmtFin SF 2',
       'Bsmt Unf SF', 'Total Bsmt SF', 'Heating', 'Heating QC', 'Central Air',
       'Electrical', '1st Flr SF', '2nd Flr SF', 'Low Qual Fin SF',
       'Gr Liv Area', 'Bsmt Full Bath', 'Bsmt Half Bath', 'Full Bath',
       'Half Bath', 'Bedroom AbvGr', 'Kitchen AbvGr', 'Kitchen Qual',
       'TotRms AbvGrd', 'Functional', 'Fireplaces', 'Fireplace Qu',
       'Garage Type', 'Garage Yr Blt', 'Garage Finish'

> There are 2930 records in the dataset and 81 features, the features comprimises of both numeric and categorical data types.

> Opening the datasets description file for better insight 

In [13]:
with open('Ames_Housing_Feature_Description.txt','r') as f: 
    print(f.read())

MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

MSZoning: Identifies the general zoning classification of the sale.
		
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM

### Wrangling /Data cleaning

> In this section I want to clean the data, deal with missing data, wrong data types and also outliers if any

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   PID              2930 non-null   int64  
 1   MS SubClass      2930 non-null   int64  
 2   MS Zoning        2930 non-null   object 
 3   Lot Frontage     2440 non-null   float64
 4   Lot Area         2930 non-null   int64  
 5   Street           2930 non-null   object 
 6   Alley            198 non-null    object 
 7   Lot Shape        2930 non-null   object 
 8   Land Contour     2930 non-null   object 
 9   Utilities        2930 non-null   object 
 10  Lot Config       2930 non-null   object 
 11  Land Slope       2930 non-null   object 
 12  Neighborhood     2930 non-null   object 
 13  Condition 1      2930 non-null   object 
 14  Condition 2      2930 non-null   object 
 15  Bldg Type        2930 non-null   object 
 16  House Style      2930 non-null   object 
 17  Overall Qual  

>From above we can see there are alot of missing data in multiple columns

In [18]:
df.head(30)

PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley  \
0   526301100           20        RL         141.0     31770   Pave   NaN   
1   526350040           20        RH          80.0     11622   Pave   NaN   
2   526351010           20        RL          81.0     14267   Pave   NaN   
3   526353030           20        RL          93.0     11160   Pave   NaN   
4   527105010           60        RL          74.0     13830   Pave   NaN   
5   527105030           60        RL          78.0      9978   Pave   NaN   
6   527127150          120        RL          41.0      4920   Pave   NaN   
7   527145080          120        RL          43.0      5005   Pave   NaN   
8   527146030          120        RL          39.0      5389   Pave   NaN   
9   527162130           60        RL          60.0      7500   Pave   NaN   
10  527163010           60        RL          75.0     10000   Pave   NaN   
11  527165230           20        RL           NaN      7980   Pave   NaN   
12  527166040           60        RL          63.0      8402   Pave   NaN   
13  527180040           20        RL          85.0     10176   Pave   NaN   
14  527182190          120        RL           NaN      6820   Pave   NaN   
15  527216070           60        RL          47.0     53504   Pave   NaN   
16  527225035           50        RL         152.0     12134   Pave   NaN   
17  527258010           20        RL          88.0     11394   Pave   NaN   
18  527276150           20        RL         140.0     19138   Pave   NaN   
19  527302110           20        RL          85.0     13175   Pave   NaN   
20  527358140           20        RL         105.0     11751   Pave   NaN   
21  527358200           85        RL          85.0     10625   Pave   NaN   
22  527368020           60        FV           NaN      7500   Pave   NaN   
23  527402200           20        RL           NaN     11241   Pave   NaN   
24  527402250           20        RL           NaN     12537   Pave   NaN   
25  527403020           20        RL          65.0      8450   Pave   NaN   
26  527404120           20        RL          70.0      8400   Pave   NaN   
27  527425090           20        RL          70.0     10500   Pave   NaN   
28  527427230          120        RH          26.0      5858   Pave   NaN   
29  527451180          160        RM          21.0      1680   Pave   NaN   

   Lot Shape Land Contour Utilities  ... Pool Area Pool QC  Fence  \
0        IR1          Lvl    AllPub  ...         0     NaN    NaN   
1        Reg          Lvl    AllPub  ...         0     NaN  MnPrv   
2        IR1          Lvl    AllPub  ...         0     NaN    NaN   
3        Reg          Lvl    AllPub  ...         0     NaN    NaN   
4        IR1          Lvl    AllPub  ...         0     NaN  MnPrv   
5        IR1          Lvl    AllPub  ...         0     NaN    NaN   
6        Reg          Lvl    AllPub  ...         0     NaN    NaN   
7        IR1          HLS    AllPub  ...         0     NaN    NaN   
8        IR1          Lvl    AllPub  ...         0     NaN    NaN   
9        Reg          Lvl    AllPub  ...         0     NaN    NaN   
10       IR1          Lvl    AllPub  ...         0     NaN    NaN   
11       IR1          Lvl    AllPub  ...         0     NaN  GdPrv   
12       IR1          Lvl    AllPub  ...         0     NaN    NaN   
13       Reg          Lvl    AllPub  ...         0     NaN    NaN   
14       IR1          Lvl    AllPub  ...         0     NaN    NaN   
15       IR2          HLS    AllPub  ...         0     NaN    NaN   
16       IR1          Bnk    AllPub  ...         0     NaN    NaN   
17       Reg          Lvl    AllPub  ...         0     NaN    NaN   
18       Reg          Lvl    AllPub  ...         0     NaN    NaN   
19       Reg          Lvl    AllPub  ...         0     NaN  MnPrv   
20       IR1          Lvl    AllPub  ...         0     NaN  MnPrv   
21       Reg          Lvl    AllPub  ...         0     NaN  MnPrv   
22       Reg          Lvl    AllPub

### Removing the PID

We already have an index, so we don't need the PID unique identifier for the regression we will perform later on.

In [7]:
df = df.drop('PID',axis=1)

In [15]:
df.isnull().sum()

PID                 0
MS SubClass         0
MS Zoning           0
Lot Frontage      490
Lot Area            0
                 ... 
Mo Sold             0
Yr Sold             0
Sale Type           0
Sale Condition      0
SalePrice           0
Length: 81, dtype: int64

>checking percentages of missing data for columns using a function to make things easier

In [19]:
#function for returning missing percentage
def percent_missing(df):
    percent_nan = 100* df.isnull().sum() / len(df)
    percent_nan = percent_nan[percent_nan>0].sort_values()
    return percent_nan

In [20]:
percent_missing(df)

Electrical         0.034130
Garage Area        0.034130
Total Bsmt SF      0.034130
Bsmt Unf SF        0.034130
BsmtFin SF 1       0.034130
BsmtFin SF 2       0.034130
Garage Cars        0.034130
Bsmt Full Bath     0.068259
Bsmt Half Bath     0.068259
Mas Vnr Area       0.784983
Mas Vnr Type       0.784983
BsmtFin Type 1     2.730375
Bsmt Qual          2.730375
Bsmt Cond          2.730375
BsmtFin Type 2     2.764505
Bsmt Exposure      2.832765
Garage Type        5.358362
Garage Finish      5.426621
Garage Qual        5.426621
Garage Cond        5.426621
Garage Yr Blt      5.426621
Lot Frontage      16.723549
Fireplace Qu      48.532423
Fence             80.477816
Alley             93.242321
Misc Feature      96.382253
Pool QC           99.556314
dtype: float64

>Dropping columns with more 50 percent of the features missing

In [21]:
df = df.drop(['Pool QC','Misc Feature','Alley','Fence'],axis=1)

In [22]:
percent_missing(df)

Total Bsmt SF      0.034130
Garage Cars        0.034130
BsmtFin SF 1       0.034130
BsmtFin SF 2       0.034130
Bsmt Unf SF        0.034130
Garage Area        0.034130
Electrical         0.034130
Bsmt Half Bath     0.068259
Bsmt Full Bath     0.068259
Mas Vnr Type       0.784983
Mas Vnr Area       0.784983
Bsmt Qual          2.730375
Bsmt Cond          2.730375
BsmtFin Type 1     2.730375
BsmtFin Type 2     2.764505
Bsmt Exposure      2.832765
Garage Type        5.358362
Garage Finish      5.426621
Garage Cond        5.426621
Garage Qual        5.426621
Garage Yr Blt      5.426621
Lot Frontage      16.723549
Fireplace Qu      48.532423
dtype: float64

>going through the description text i have decided to fill the null values of some of the categorical columns as "None" instead of dropping them, as some houses have no features such as basements, fireplace, garages etc

In [23]:
cat_cols =  ['Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin Type 2','Mas Vnr Type','Garage Type',
                  'Garage Finish', 'Garage Qual', 'Garage Cond','Fireplace Qu']
df[cat_cols] = df[cat_cols].fillna('None')

In [24]:
percent_missing(df)

BsmtFin SF 1       0.034130
BsmtFin SF 2       0.034130
Bsmt Unf SF        0.034130
Total Bsmt SF      0.034130
Electrical         0.034130
Garage Cars        0.034130
Garage Area        0.034130
Bsmt Full Bath     0.068259
Bsmt Half Bath     0.068259
Mas Vnr Area       0.784983
Garage Yr Blt      5.426621
Lot Frontage      16.723549
dtype: float64

>Doing the same for a few numerical columns assuming they dont have associated properties such as basements; No basements means no basement measurements

In [25]:
num_cols = ['BsmtFin SF 1', 'BsmtFin SF 2', 'Bsmt Unf SF','Total Bsmt SF', 'Bsmt Full Bath', 'Bsmt Half Bath',
            'Mas Vnr Area','Garage Yr Blt','Lot Frontage']
df[num_cols] = df[num_cols].fillna(0)

In [26]:
percent_missing(df)

Electrical     0.03413
Garage Cars    0.03413
Garage Area    0.03413
dtype: float64

>dropping the remaining rows as they are minimal and wont affect performance of the models

In [27]:
df.dropna(axis=0,inplace=True)

In [28]:
percent_missing(df)

Series([], dtype: float64)

>all missing data have been dealt with appropriately now moving on to vizualisations dealing with outliers if any

In [29]:
pwd

'C:\\Users\\David\\Desktop\\Project1s\\Ames Data Housing'